{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Finder\n",
    "The purpose of this function is to return genbank genome data as an organized `.csv` file with relevant data. \n",
    "This data includes chromosome location, name, locus tag, start codon location, stop codon location, \n",
    "direction of transcription, and the entire gene sequence. Listing genomic data in this fashion improves\n",
    "bioinformatics related workflows where gene sequences are a focus. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "import csv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def gene_finder(file, chromosome = np.NaN):\n",
    "    chromosome_number = chromosome\n",
    "    filename = file\n",
    "    genes = pd.DataFrame(columns=['chromosome', 'gene names', 'locus tag', 'start', 'stop', 'direction', 'sequence'])\n",
    "    record = SeqIO.read(filename, \"genbank\")\n",
    "    currentChromosome = record.seq\n",
    "    sequence = \"\"\n",
    "    name = ''\n",
    "    locus = ''\n",
    "    start = ''\n",
    "    stop = ''\n",
    "    direction = ''\n",
    "    sequences = ''\n",
    "    \n",
    "    for feature in record.features:\n",
    "        if(feature.type == \"gene\"):\n",
    "            # gene name\n",
    "            name = str(feature.qualifiers.get(\"gene\"))    \n",
    "        if(feature.type == \"gene\"):\n",
    "            # locus tag (unique)\n",
    "            locus =  str(feature.qualifiers.get(\"locus_tag\"))\n",
    "        if(feature.type == \"gene\"):\n",
    "            # start codon\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace(\"[\", \"\")\n",
    "            edit2 = edit1.replace(\"<\",\"\")\n",
    "            edit3 = edit2.replace(\">\",\"\")\n",
    "            edit4 = edit3.replace(\"]\",\"\")\n",
    "            edit5 = edit4.replace(\"(-)\",\"\")\n",
    "            edit6 = edit5.replace(\"(+)\",\"\")\n",
    "            splitWords = edit6.split(\":\")\n",
    "            start = str(splitWords[0])\n",
    "        if(feature.type == \"gene\"):\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace(\"[\", \"\")\n",
    "            edit2 = edit1.replace(\"<\",\"\")\n",
    "            edit3 = edit2.replace(\">\",\"\")\n",
    "            edit4 = edit3.replace(\"]\",\"\")\n",
    "            edit5 = edit4.replace(\"(-)\",\"\")\n",
    "            edit6 = edit5.replace(\"(+)\",\"\")\n",
    "            splitWords = edit6.split(\":\")\n",
    "            stop = str(splitWords[1])\n",
    "        if(feature.type == \"gene\"):\n",
    "            original_location = str(feature.location)\n",
    "            splitWords = original_location.split(\"]\")\n",
    "            direction = str(splitWords[1])\n",
    "        if(feature.type==\"gene\"):\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace(\"[\", \"\")\n",
    "            edit2 = edit1.replace(\"<\",\"\")\n",
    "            edit3 = edit2.replace(\">\",\"\")\n",
    "            edit4 = edit3.replace(\"]\",\"\")\n",
    "            edit5 = edit4.replace(\"(-)\",\"\")\n",
    "            edit6 = edit5.replace(\"(+)\",\"\")\n",
    "            splitWords = edit6.split(\":\")\n",
    "            lower = int(splitWords[0])\n",
    "            upper = int(splitWords[1])\n",
    "            sequence = currentChromosome[lower:upper]\n",
    "            sequences = sequence\n",
    "        genes = genes.append({'chromosome': chromosome_number, 'gene names': name, 'locus tag': locus, \n",
    "        'start': start, 'stop': stop, 'direction': direction, 'sequence': sequences}, ignore_index=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "import csv\n",
    "import numpy\n",
    "import pandas as pd\n",
    "\n",
    "filename = \"chromosomes/chromosome 0.flat\"\n",
    "df = pd.DataFrame(columns=['chromosome', 'gene names', 'locus tag', 'start', 'stop', 'direction', 'sequence'])\n",
    "\n",
    "for i in range(15):\n",
    "    filenumber = i + 1\n",
    "    filename = filename.replace(str(i), str(i + 1))\n",
    "    record = SeqIO.read(filename, \"genbank\")\n",
    "    currentChromosome = record.seq\n",
    "    sequence = \"\"\n",
    "    name = ''\n",
    "    locus = ''\n",
    "    start = ''\n",
    "    stop = ''\n",
    "    direction = ''\n",
    "    sequences = ''\n",
    "    \n",
    "    for feature in record.features:\n",
    "        if(feature.type == \"gene\"):\n",
    "            # gene name\n",
    "            name = str(feature.qualifiers.get(\"gene\"))    \n",
    "        if(feature.type == \"gene\"):\n",
    "            # locus tag (unique)\n",
    "            locus =  str(feature.qualifiers.get(\"locus_tag\"))\n",
    "        if(feature.type == \"gene\"):\n",
    "            # start codon\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace(\"[\", \"\")\n",
    "            edit2 = edit1.replace(\"<\",\"\")\n",
    "            edit3 = edit2.replace(\">\",\"\")\n",
    "            edit4 = edit3.replace(\"]\",\"\")\n",
    "            edit5 = edit4.replace(\"(-)\",\"\")\n",
    "            edit6 = edit5.replace(\"(+)\",\"\")\n",
    "            splitWords = edit6.split(\":\")\n",
    "            start = str(splitWords[0])\n",
    "        if(feature.type == \"gene\"):\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace(\"[\", \"\")\n",
    "            edit2 = edit1.replace(\"<\",\"\")\n",
    "            edit3 = edit2.replace(\">\",\"\")\n",
    "            edit4 = edit3.replace(\"]\",\"\")\n",
    "            edit5 = edit4.replace(\"(-)\",\"\")\n",
    "            edit6 = edit5.replace(\"(+)\",\"\")\n",
    "            splitWords = edit6.split(\":\")\n",
    "            stop = str(splitWords[1])\n",
    "        if(feature.type == \"gene\"):\n",
    "            original_location = str(feature.location)\n",
    "            splitWords = original_location.split(\"]\")\n",
    "            direction = str(splitWords[1])\n",
    "        if(feature.type==\"gene\"):\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace(\"[\", \"\")\n",
    "            edit2 = edit1.replace(\"<\",\"\")\n",
    "            edit3 = edit2.replace(\">\",\"\")\n",
    "            edit4 = edit3.replace(\"]\",\"\")\n",
    "            edit5 = edit4.replace(\"(-)\",\"\")\n",
    "            edit6 = edit5.replace(\"(+)\",\"\")\n",
    "            splitWords = edit6.split(\":\")\n",
    "            lower = int(splitWords[0])\n",
    "            upper = int(splitWords[1])\n",
    "            sequence = currentChromosome[lower:upper]\n",
    "            sequences = sequence\n",
    "        df = df.append({'chromosome': filenumber, 'gene names': name, 'locus tag': locus, \n",
    "        'start': start, 'stop': stop, 'direction': direction, 'sequence': sequences}, ignore_index=True)\n",
    "\n",
    "df.drop_duplicates(subset = \"locus tag\", keep = \"first\", inplace = True)\n",
    "df.dropna(subset = [\"locus tag\"], inplace=True)\n",
    "\n",
    "df.to_csv(\"chromosomes.csv\", index = False)\n",
    "print(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.8.8 (default, Apr 13 2021, 12:59:45) \n[Clang 10.0.0 ]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
