{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Finder\n",
    "The purpose of this function is to return genbank genome data as an organized pandas dataframe with relevant data. \n",
    "This data includes chromosome location, name, locus tag, start codon location, stop codon location, \n",
    "direction of transcription, and the entire gene sequence. Listing genomic data in this fashion improves\n",
    "bioinformatics related workflows where gene sequences are a focus. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "import csv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "def gene_finder(file, chromosome = np.NaN):\n",
    "    chromosome_number = chromosome\n",
    "    filename = file\n",
    "    genes = pd.DataFrame(columns=['chromosome', 'gene name', 'locus tag', 'start', 'stop', 'direction', 'sequence'])\n",
    "    record = SeqIO.read(filename, 'genbank')\n",
    "    currentChromosome = record.seq\n",
    "    sequence = ''\n",
    "    name = ''\n",
    "    locus = ''\n",
    "    start = ''\n",
    "    stop = ''\n",
    "    direction = ''\n",
    "    sequences = ''\n",
    "    \n",
    "    for feature in record.features:\n",
    "        if(feature.type == 'gene'):\n",
    "            # gene name\n",
    "            name = str(feature.qualifiers.get('gene'))    \n",
    "        if(feature.type == 'gene'):\n",
    "            # locus tag (unique)\n",
    "            locus =  str(feature.qualifiers.get('locus_tag'))\n",
    "        if(feature.type == 'gene'):\n",
    "            # start codon\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace('[', '')\n",
    "            edit2 = edit1.replace('<','')\n",
    "            edit3 = edit2.replace('>','')\n",
    "            edit4 = edit3.replace(']','')\n",
    "            edit5 = edit4.replace('(-)','')\n",
    "            edit6 = edit5.replace('(+)','')\n",
    "            splitWords = edit6.split(':')\n",
    "            start = str(splitWords[0])\n",
    "        if(feature.type == 'gene'):\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace('[', '')\n",
    "            edit2 = edit1.replace('<','')\n",
    "            edit3 = edit2.replace('>','')\n",
    "            edit4 = edit3.replace(']','')\n",
    "            edit5 = edit4.replace('(-)','')\n",
    "            edit6 = edit5.replace('(+)','')\n",
    "            splitWords = edit6.split(':')\n",
    "            stop = str(splitWords[1])\n",
    "        if(feature.type == 'gene'):\n",
    "            original_location = str(feature.location)\n",
    "            splitWords = original_location.split(']')\n",
    "            direction = str(splitWords[1])\n",
    "        if(feature.type=='gene'):\n",
    "            original_location = str(feature.location)\n",
    "            edit1 = original_location.replace('[', '')\n",
    "            edit2 = edit1.replace('<','')\n",
    "            edit3 = edit2.replace('>','')\n",
    "            edit4 = edit3.replace(']','')\n",
    "            edit5 = edit4.replace('(-)','')\n",
    "            edit6 = edit5.replace('(+)','')\n",
    "            splitWords = edit6.split(':')\n",
    "            lower = int(splitWords[0])\n",
    "            upper = int(splitWords[1])\n",
    "            sequence = currentChromosome[lower:upper]\n",
    "            sequences = sequence\n",
    "        genes = genes.append({'chromosome': chromosome_number, 'gene name': name, 'locus tag': locus, \n",
    "        'start': start, 'stop': stop, 'direction': direction, 'sequence': sequences}, ignore_index=True)\n",
    "    \n",
    "    genes.drop_duplicates(subset = 'locus tag', keep = 'first', inplace = True)\n",
    "    genes.dropna(subset = ['locus tag'], inplace=True)\n",
    "    genes.drop(0, inplace = True)\n",
    "    genes.reset_index(inplace = True)\n",
    "    genes.drop(columns='index', inplace = True)\n",
    "    genes['sequence'] = genes['sequence'].astype(str)\n",
    "    return genes \n",
    "    "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is an example of this function in action with the data from the first chromosome of \n",
    "Saccharomyces cerevisiae. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene name</th>\n",
       "      <th>locus tag</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>direction</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>['PAU8']</td>\n",
       "      <td>['YAL068C']</td>\n",
       "      <td>1806</td>\n",
       "      <td>2169</td>\n",
       "      <td>(-)</td>\n",
       "      <td>CTAGTTTGCGATAGTGTAGATACCGTCCTTGGATAGAGCACTGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAL067W-A']</td>\n",
       "      <td>2479</td>\n",
       "      <td>2707</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGCCAATTATAGGGGTGCCGAGGTGCCTTATAAAACCCTTTTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>['SEO1']</td>\n",
       "      <td>['YAL067C']</td>\n",
       "      <td>7234</td>\n",
       "      <td>9016</td>\n",
       "      <td>(-)</td>\n",
       "      <td>TTATTTTTCATCAGATACTGATAAGGTTTCAACGTCTTTTGACGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAL065C']</td>\n",
       "      <td>11564</td>\n",
       "      <td>11951</td>\n",
       "      <td>(-)</td>\n",
       "      <td>TTACCATACGATTGCCAGCAATACGGTGGAAATAAAAACACTTATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAL064W-B']</td>\n",
       "      <td>12045</td>\n",
       "      <td>12426</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGGCAGGTGAAGCAGTTTCGGAACACACACCAGATTCGCAGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAR061W']</td>\n",
       "      <td>218139</td>\n",
       "      <td>219145</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGCCTTATCACTATTTATTTTTGGCACTCTTCACCTACCTGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAR064W']</td>\n",
       "      <td>220197</td>\n",
       "      <td>220497</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGTTGATTGATTTTTGCTGTAGTTATATAGCAGGGACCCACGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAR066W']</td>\n",
       "      <td>221048</td>\n",
       "      <td>221660</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGTTCAATCGTTTTAACAAATTCCAAGCTGCTGTCGCTTTGGCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>['YAR068W']</td>\n",
       "      <td>222405</td>\n",
       "      <td>222891</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGCCACAAGTACAGTCGTGGTTTCCTGTTCAGAAACAACCGACGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>1</td>\n",
       "      <td>['PHO11']</td>\n",
       "      <td>['YAR071W']</td>\n",
       "      <td>225459</td>\n",
       "      <td>226863</td>\n",
       "      <td>(+)</td>\n",
       "      <td>ATGTTGAAGTCAGCCGTTTATTCAATTTTAGCCGCTTCTTTGGTTA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome  gene name      locus tag   start    stop direction  \\\n",
       "0            1   ['PAU8']    ['YAL068C']    1806    2169       (-)   \n",
       "1            1       None  ['YAL067W-A']    2479    2707       (+)   \n",
       "2            1   ['SEO1']    ['YAL067C']    7234    9016       (-)   \n",
       "3            1       None    ['YAL065C']   11564   11951       (-)   \n",
       "4            1       None  ['YAL064W-B']   12045   12426       (+)   \n",
       "..         ...        ...            ...     ...     ...       ...   \n",
       "96           1       None    ['YAR061W']  218139  219145       (+)   \n",
       "97           1       None    ['YAR064W']  220197  220497       (+)   \n",
       "98           1       None    ['YAR066W']  221048  221660       (+)   \n",
       "99           1       None    ['YAR068W']  222405  222891       (+)   \n",
       "100          1  ['PHO11']    ['YAR071W']  225459  226863       (+)   \n",
       "\n",
       "                                              sequence  \n",
       "0    CTAGTTTGCGATAGTGTAGATACCGTCCTTGGATAGAGCACTGGAG...  \n",
       "1    ATGCCAATTATAGGGGTGCCGAGGTGCCTTATAAAACCCTTTTCTG...  \n",
       "2    TTATTTTTCATCAGATACTGATAAGGTTTCAACGTCTTTTGACGTT...  \n",
       "3    TTACCATACGATTGCCAGCAATACGGTGGAAATAAAAACACTTATG...  \n",
       "4    ATGGCAGGTGAAGCAGTTTCGGAACACACACCAGATTCGCAGGAAG...  \n",
       "..                                                 ...  \n",
       "96   ATGCCTTATCACTATTTATTTTTGGCACTCTTCACCTACCTGGCCA...  \n",
       "97   ATGTTGATTGATTTTTGCTGTAGTTATATAGCAGGGACCCACGGAA...  \n",
       "98   ATGTTCAATCGTTTTAACAAATTCCAAGCTGCTGTCGCTTTGGCCC...  \n",
       "99   ATGCCACAAGTACAGTCGTGGTTTCCTGTTCAGAAACAACCGACGC...  \n",
       "100  ATGTTGAAGTCAGCCGTTTATTCAATTTTAGCCGCTTCTTTGGTTA...  \n",
       "\n",
       "[101 rows x 7 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_1 = gene_finder('assets/chromosome 1.flat', 1)\n",
    "chr_1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data can easily be exported as a `.csv` or Microsoft Excel file with pandas. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export to .csv\n",
    "chr_1.to_csv('assets/chr_1.csv', index = False)\n",
    "\n",
    "# export to .xlsx (Microsoft Excel)\n",
    "chr_1.to_excel('assets/chr_1.xlsx', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8 (default, Apr 13 2021, 12:59:45) \n[Clang 10.0.0 ]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "40d3a090f54c6569ab1632332b64b2c03c39dcf918b08424e98f38b5ae0af88f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
